scio-tensorflow/src/main/scala/com/spotify/scio/tensorflow/TensorFlowDoFns.scala

```diff
@@ -111,16 +106,25 @@ sealed trait PredictDoFn[T, V, M <: Model[_]]
 
   @Teardown
   override def teardown(): Unit = {
-    Log.info(s"Tearing down predict DoFn $this")
-    val (running, m) = getResource().get(modelId)
-    if (running.decrementAndGet() == 0) {
-      m.close()
-    }
+    Log.info("Tearing down predict DoFn {}", this)
+    getResource.compute(
+      modelId,
+      {
+        case (_, null) =>
+          Log.warn("No model to close while tearing down predict DoFn")
+          null
+        case (_, (1, model)) =>
+          model.close()
```

Comment:
This `close()` is the real side effect, and in a map with a non-locking `compute` it would be a problem because it would lead to a retry trying to close an already closed model. But in this case it works because it is a `ConcurrentHashMap`. It is good to have some comments to explain this implementation detail since it's hard to enforce with unit tests.

I can add it since it is technically my PR.